src/main/java/core/basesyntax/MyLinkedList.java

```diff
@@ -3,45 +3,219 @@
 import java.util.List;
 
 public class MyLinkedList<T> implements MyLinkedListInterface<T> {
+    private int size;
+    private Node<T> head;
+    private Node<T> tail;
+
+    static class Node<T> {
+        private T value;
+        private Node<T> prev;
+        private Node<T> next;
+
+        public Node(Node<T> prev, T value, Node<T> next) {
+            this.value = value;
+            this.prev = prev;
+            this.next = next;
+        }
+    }
+
+    public void linkLast(T value) {
+        Node<T> newNode = new Node<>(tail, value, null);
+        if (head == null) {
+            head = newNode;
+            tail = newNode;
+        } else {
+            tail.next = newNode;
+            tail = newNode;
+        }
+        size++;
+    }
+
+    public void linkFirst(T value) {
+        Node<T> newNode = new Node<>(null, value, head);
+        if (head == null) {
+            head = newNode;
+            tail = newNode;
+        } else {
+            head.prev = newNode;
+            head = newNode;
+        }
+        size++;
+    }
+
+    public void linkByIndex(T value, int index) {
+        indexCheck(index);
+
+        if (index == 0) {
+            linkFirst(value);
+        } else if (index == size) {
+            linkLast(value);
+        } else {
+            Node<T> nodeByIndex = findNodeByIndex(index);
+            Node<T> newNode = new Node<>(nodeByIndex.prev, value, nodeByIndex);
+            nodeByIndex.prev.next = newNode;
+            nodeByIndex.prev = newNode;
+            size++;
+        }
+    }
+
+    public Node<T> findNodeByIndex(int index) {
+        indexCheck(index);
+        Node<T> currentNode;
+
+        if (index < size >> 1) {
+            currentNode = head;
+            for (int i = 0; i < index; i++) {
+                currentNode = currentNode.next;
+            }
+        } else {
+            currentNode = tail;
+            for (int i = size - 1; i > index; i--) {
+                currentNode = currentNode.prev;
+            }
+        }
+        return currentNode;
+    }
+
+    public void indexCheck(int index) {
+        if (index < 0 || index > size) {
+            throw new IndexOutOfBoundsException(
+                    "indexCheckException: Wrong index < 0 || index > size");
```

Comment:
The `indexCheck` method message in the exception could be more informative by including the actual index and size values to help with debugging.